refactor(plugin-odap-gateway): refactor implementation including more tests

AndreAugusto11 opened the pull request:
Refactor first implementation #1275 

* Rename plugin file from **odap-gateway** to **plugin-odap-gateway**.
* Refactor **SendClientRequest()** 600 lines method into different methods according to the ODAP phase and each message type.
* Refactor methods that make digital signatures and respective verification.
* Added tests and actual verification regarding the state of the objects.
* Added fields in the open API file according to the specification in the ODAP draft.
* Added checks in every step of the protocol according to the messages exchanged

cc @RafaelAPB 

Signed-off-by: André Augusto <andre.augusto@tecnico.ulisboa.pt>

takeutak commented:
@AndreAugusto11 It is good for me, but would you add an issue corresponding to this pull-request? (I'm afraid that this is a formal indication.)
I think the content should be based on your following description, so please make it referring to past issues.

> Rename plugin file from odap-gateway to plugin-odap-gateway.
> Refactor SendClientRequest() 600 lines method into different methods according to the ODAP phase and each message type.
> Refactor methods that make digital signatures and respective verification.
> Added tests and actual verification regarding the state of the objects.
> Added fields in the open API file according to the specification in the ODAP draft.
> Added checks in every step of the protocol according to the messages exchanged

RafaelAPB commented:
@petermetz the ci fails, looks something related to the fabric test ledger? https://github.com/hyperledger/cactus/runs/5470188361?check_suite_focus=true#step:8:22867

AndreAugusto11 commented:
It also throws an error when running the **_odap-api-call.test.ts_**, and there are no tests found in that file, don't know why.
https://github.com/hyperledger/cactus/runs/5470188361?check_suite_focus=true#step:8:35844

jscode017 commented:
@RafaelAPB @AndreAugusto11 
There may be something I did not merge from my evaluation branch
Could you try adding 
`  public fabricAssetSize?: string;`
in the OdapGatewayConstructorOptions
and
`if (!options.fabricAssetSize) {
    this.fabricAssetSize = "1";
} else {
  this.fabricAssetSize = options.fabricAssetSize;
}`
in the odap gateway constructor
To see if it fix the problem
 sorry, I should create a pr recently, but kind of busy doing school coursework and RA

RafaelAPB commented:
@AndreAugusto11 the test passes locally with JEST. Perhaps the test runner is using TAP. Could you double check that the test runner that runs that test is JEST?

RafaelAPB commented:
@jscode017 thank you for the suggestion.

AndreAugusto11 commented:
@RafaelAPB I think you're right, all the tests using Jest have passed and then in line [35748](https://github.com/hyperledger/cactus/runs/5470188361?check_suite_focus=true#step:8:35748) it initiates TAP version 13. Afterwards, the test fails. I already removed this test from the **jest.config.js** file. Any suggestion why it is being ignored by Jest?

petermetz commented on jest.config.js:
```diff
@@ -11,11 +11,7 @@ module.exports = {
   // Ignore the tests that are still using tap/tape for as their test runner
   testPathIgnorePatterns: [
     `./packages/cactus-plugin-keychain-aws-sm/src/test/typescript/integration/openapi/openapi-validation.test.ts`,
```
@AndreAugusto11 You also have to make these same ignore/include pattern changes in the .taprc file so that the tests that are now using Jest will not be executed by tap/tape. (so head over to that file and remove these files from the include list is what you need to do if I remember correctly).

Once that's sorted out the CI should start passing, assuming that the tests are all working fine otherwise.

AndreAugusto11 replied:
@petermetz that was what I was missing! I didn't know about that config file. It is now passing all checks. TY

petermetz commented:
> @RafaelAPB @AndreAugusto11 There may be something I did not merge from my evaluation branch Could you try adding ` public fabricAssetSize?: string;` in the OdapGatewayConstructorOptions and `if (!options.fabricAssetSize) { this.fabricAssetSize = "1"; } else { this.fabricAssetSize = options.fabricAssetSize; }` in the odap gateway constructor and change this from number to string like: https://github.com/jscode017/cactus/blob/552a5b0801896b2894f678fc8c9414496bbd12ca/packages/cactus-plugin-odap-hermes/src/main/json/openapi.json#L433 To see if it fix the problem sorry, I should create a pr recently, but kind of busy doing school coursework and RA

@AndreAugusto11 If you can confirm that these changes were made as well then we are good to go from my side, just let me know explicitly please, I wasn't sure if the thumbs up on Jason's comment was meant to imply that or not. :-)

AndreAugusto11 commented:
> @AndreAugusto11 If you can confirm that these changes were made as well then we are good to go from my side, just let me know explicitly please, I wasn't sure if the thumbs up on Jason's comment was meant to imply that or not. :-)
Hey @petermetz, my bad, I already made those changes. I've just also updated the docs of the package and rebased, so it is ready to merge into main once the ci is complete.

petermetz reviewed (approved):
@AndreAugusto11 No worries, thank you for confirming! LGTM

RafaelAPB commented:
LGMT too :)

petermetz reviewed (changes requested):
@AndreAugusto11 Please fix the merge conflicts and then we are good to go

AndreAugusto11 commented:
@petermetz Done :)

petermetz reviewed (approved):
@AndreAugusto11 Great! Thank you very much!

petermetz commented:
@AndreAugusto11 I just noticed that the commit message is pretty much empty compared to the PR description. Next time please include the same information in the commit message.

AndreAugusto11 commented:
@petermetz you're right, I could've added more information to the commit message as well. I'm sorry for that, I'll take it into account the next time.

petermetz commented:
@AndreAugusto11 No worries at all, I try not to nit pick on everything every single time (even if that's hard to believe haha). Thanks for your understanding! :-)
